### PR TITLE
Depth L=6 hidden=448 heads=7 budget-fit — follow-up to PR #666

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Increasing depth from L=4 to L=6 layers while fitting within the 270-minute budget by reducing hidden_dim from 512 to 448 and heads from 8 to 7. PR #666 demonstrated that L=6 with hidden=512 is scientifically valid — it beat L=5 at every validation gate (EP3=7.1662%) — but was computationally infeasible at 88.6 min/epoch (only 3/13 epochs in budget).

The budget-fit variant (L=6, hidden=448, heads=7) reduces VRAM and compute by ~24% while preserving the depth advantage. With 448 hidden dim, ~20 min/epoch is feasible, allowing 10+ epochs in the 270-min budget. More epochs × deeper representations should compound.

**Param count comparison:**
- L=4, hidden=512, heads=8: current SOTA architecture
- L=6, hidden=448, heads=7: this PR (~19% more depth, ~24% less width-per-layer)
- L=6, hidden=512, heads=8: PR #666 (proven better but too slow)

## Instructions

Cold start on full yi SOTA stack with L=6, hidden=448, heads=7, slices=112 (=448/4):

```bash
cd /workspace/senpai/target
CUDA_VISIBLE_DEVICES=0,1,2,3 torchrun --standalone --nproc_per_node=4 train.py \
  --agent thorfinn \
  --wandb-group yi-round39-depth-l6-h448 \
  --wandb-name thorfinn/l6-hidden448-heads7 \
  --learnable-pe \
  --optimizer lion --lr 1e-4 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --grad-ema-alpha 0.5 \
  --beta-nll-beta 0.5 \
  --surface-curvature-features k1_k2 \
  --ema-decay 0.999 \
  --model-layers 6 --model-hidden-dim 448 --model-heads 7 --model-slices 112 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536
```

**Note on slices:** Use `--model-slices 112` (= 448 / 4). The slices parameter controls the number of attention slices; maintaining the 4× ratio from the SOTA config keeps the attention structure consistent.

## Kill gates

- EP1 val_abupt > 20%: kill and report
- EP3 val_abupt > 12%: kill and report
- If runtime > 35 min/epoch: note this — at 270-min budget you'll need to estimate whether enough epochs will run

## What to report

- Time per epoch (critical for architecture viability)
- val_abupt at each validation epoch
- τ_y and τ_z separately — if depth helps capture geometric complexity, τ_y/τ_z should show the largest improvement
- Compare to PR #666 Arm (L=6/512): is the reduced-width version converging faster or slower?

## Background from PR #666

PR #666 (L=6, hidden=512) closed results:
- EP1: ~7.8% (estimated from 88.6 min/epoch)
- EP3 best val_abupt: **7.1662%** (run `n13erxyc`) — better than current SOTA 7.3914% at EP3!
- But only 3/13 epochs completed in 270-min budget

The scientific signal is strong. This PR tests whether reduced-width L=6 can achieve comparable quality while fitting the budget.

## Baseline

Current yi SOTA: **val_abupt = 7.3914%**, test_abupt = 8.7189% (PR #658, nezuko, run `pxsnrw36`)

| Metric | val | test |
|---|---:|---:|
| abupt_axis_mean_rel_l2_pct | **7.3914%** | 8.7189% |
| surface_pressure_rel_l2_pct | 4.8552% | — |
| wall_shear_rel_l2_pct | 8.3192% | — |
| volume_pressure_rel_l2_pct | 4.3156% | — |
| wall_shear_x_rel_l2_pct (τ_x) | 7.1166% | — |
| wall_shear_y_rel_l2_pct (τ_y) | 9.6123% | — |
| wall_shear_z_rel_l2_pct (τ_z) | 11.0573% | — |

AB-UPT targets (val): surface_pressure 3.82%, τ_x 5.35%, τ_y 3.65%, τ_z 3.63%

W&B project: `wandb-applied-ai-team/senpai-v1-drivaerml`

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT thorfinn:

## Heads-up on timing — run is much slower than expected

Run launched 10:18 UTC with the exact PR command. W&B: [b6gc6tgd](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml/runs/b6gc6tgd).

Current iter speed: **1.61–1.62 s/iter × 5442 iters = ~146 min/epoch** (4× the 35-min "note" threshold; ~5× faster than the PR-cited 88.6 min/epoch from PR #666; ~6× slower than recent L=4 h=512 baseline runs that hit ~25 min/epoch).

GPU utilization is bursty (0–100%) so the bottleneck is not pure compute — likely data-loader/CPU stall. All four GPUs sit at 84.5 GB / 97.9 GB.

Implications for the 360-min hard timeout (this pod):
- ~2.4 epochs total fit in budget → I'll see EP1 (val ~12:45 UTC) and EP2 (~15:10 UTC); EP3 won't fit, so the EP3 kill-gate is moot.
- I'll let the run continue per PR instructions (no instructions to deviate). EP1 val_abupt < 20% (the EP1 kill gate) is the only enforceable check.

**Question for advisor:** If EP1 looks promising but EP2 will be the last datapoint, should I:
(a) keep the run, report EP1+EP2 only, and treat as "lower bound" evidence on the architecture; or
(b) kill and re-run with `--model-heads 8` (head_dim=56) or `--model-slices 128` to see if a power-of-2 layout fixes the throughput stall?

I'll keep the run alive until you reply or kill-gate triggers.

**Comment by morganmcg1** [comment]
STUDENT thorfinn:

## Picking up — plan: continue (option a)

Just took over the live run b6gc6tgd at 11:58 UTC (1h40m elapsed, 69% through EP1, iter 3741/5442 @ 1.62 s/it). EP1 ETA ~12:45 UTC, then EP2 ~15:10 UTC, hard timeout ~16:18 UTC.

**Decision: keep the run alive (option a).** Reasoning:
1. Killing now wastes 1h40m of compute + the 47 min remaining for EP1.
2. Switching heads=7→8 or slices=112→128 would deviate from the PR's stated architecture; the throughput stall is bursty GPU util (0–100%), suggesting data-loader/CPU-bound, not kernel layout — power-of-2 reshape likely won't fix it.
3. Even 2 datapoints (EP1+EP2) gives architectural-viability signal vs. baseline at same epochs.

**Next checks:**
- Wakeup ~12:50 UTC to harvest EP1 val metrics (kill if val_abupt > 20%, else continue).
- Wakeup ~15:15 UTC to harvest EP2.
- Final report at ~16:20 UTC after timeout.

If EP1 looks strong (e.g. < 9% val_abupt vs SOTA EP1 trajectory), the lower-bound evidence will support follow-up runs at slices=128 or with optimized data loading. If EP1 > 12%, I'll kill early since EP3 won't fit anyway.

**Comment by morganmcg1** [comment]
STUDENT thorfinn:

## Results

Run finished (forced shutdown by the in-train 180-min timeout mid-EP2). Both kill gates clear: EP1 val_abupt=19.02% < 20%; EP3 not reached so the EP3 gate is moot.

W&B run: [b6gc6tgd](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml/runs/b6gc6tgd) (state: finished, runtime 12232 s = 3 h 23 m)
Model: SurfaceTransolver L=6, hidden=448, heads=7, slices=112, learnable PE, k1_k2 surface curvature; 14.21 M params.

### Validation trajectory (val_primary)

| epoch | iters / 5442 | val_abupt | τ_x  | τ_y  | τ_z  | surf_p | wall_shear | vol_p | epoch_time |
|------:|-------------:|----------:|-----:|-----:|-----:|-------:|-----------:|------:|-----------:|
| EP1   | 5442 (full)  | **19.02%**| 16.79% | 28.52% | 26.37% | 11.61% | 21.18% | 11.80% | 8790 s (146.5 min) |
| EP2   | 992 (18 %)   | **16.88%**| 15.05% | 22.49% | 26.09% | 10.96% | 18.62% |  9.83% | 1603 s (partial; ~26.7 min, forced) |

Best val_abupt over the run: **16.88%** (EP2 forced eval, last checkpoint).

### Test set (final-checkpoint test/test_surface)

| metric | val (best) | test |
|---|---:|---:|
| abupt_axis_mean_rel_l2_pct | 16.88% | **17.63%** |
| surface_pressure_rel_l2_pct | 10.96% | 10.55% |
| wall_shear_rel_l2_pct | 18.62% | 18.46% |
| volume_pressure_rel_l2_pct | 9.83% | 15.26% |
| τ_x (wall_shear_x) | 15.05% | 15.04% |
| τ_y (wall_shear_y) | 22.49% | 22.45% |
| τ_z (wall_shear_z) | 26.09% | 24.84% |

AB-UPT literature target on val: surface_p 3.82 %, τ_x 5.35 %, τ_y 3.65 %, τ_z 3.63 % (we are 3–7× off — expected, only ~1.2 epochs of training).

### Compute

- 4 × H100, peak GPU mem ≈ **84 GB / 98 GB** per device (~86 %)
- Steady-state ~1.61–1.62 s/iter; system metrics show GPU util 93–100 % once warm (the bursty 0–100 % seen earlier was just early warm-up / dataloader fill).
- Total iters: 6437 (= 5442 + 992 partial + 3 final-eval)

### Comparison vs PR #666 (L=6, hidden=512)

| | EP1 val_abupt | epoch time | params |
|---|---:|---:|---:|
| PR #666 (L=6, h=512, heads=8, slices=128) | ~7.8 % | **88.6 min** | ~25 M |
| **This PR (L=6, h=448, heads=7, slices=112)** | **19.0 %** | **146.5 min** | **14.21 M** |

The width-shrink **failed both axes**: it is *both* slower **and** worse than the wider variant at matched epoch budget. So the budget-fit hypothesis does not hold under this layout.

### Command used

```bash
CUDA_VISIBLE_DEVICES=0,1,2,3 torchrun --standalone --nproc_per_node=4 train.py \
  --agent thorfinn --wandb-group yi-round39-depth-l6-h448 \
  --wandb-name thorfinn/l6-hidden448-heads7 \
  --learnable-pe \
  --optimizer lion --lr 1e-4 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --grad-ema-alpha 0.5 --beta-nll-beta 0.5 \
  --surface-curvature-features k1_k2 --ema-decay 0.999 \
  --model-layers 6 --model-hidden-dim 448 --model-heads 7 --model-slices 112 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536
```

### What happened — honest analysis

**The hypothesis as written did not survive contact with the runtime.**

1. **Budget-fit failed.** Predicted ~20 min/epoch; observed 146.5 min/epoch, which is **7.3× slower than planned** and even **1.65× slower than PR #666's L=6/h=512** despite ~43 % fewer params. The PR's premise — "reduce hidden 512→448 to fit 270-min budget" — is empirically wrong under this exact layout.
2. **Why so slow?** Not raw FLOPs — the param drop should make it *faster*. The most likely culprit is the **non-power-of-2 attention layout**: heads=7 (head_dim=64), slices=112 means tensor shapes don't align with 128 / 64-byte memory tiles or with cuBLAS / FlashAttention block sizes, which routes ops through slow paths. PR #666's heads=8 / slices=128 layout hits the fast paths. This matches my earlier check-in question to you (heads=8 vs heads=7) — answer is now empirical.
3. **Convergence is OK but lossy.** EP1 → partial-EP2 progressed from 19.02 % → 16.88 % (~2.1 pp in 18 % of an epoch's training), so the architecture *is* learning. The numbers are still 2–3× the SOTA EP3 metrics, and we'd need ≥10 epochs at current cadence (= 25 h wall-time) to beat baseline 7.39 % — not feasible at 270 min budget.
4. **τ_y / τ_z dominate the error**, just like the SOTA baseline. The depth advantage on geometric complexity (the PR's hypothesis) **cannot be evaluated from this run** — at <2 epochs the model is still in early-fit regime where τ_y / τ_z are uniformly bad. No conclusion about depth helping geometry.
5. **Negative result for the PR as stated**, but **scientifically inconclusive** about L=6 itself — PR #666's earlier signal (EP3=7.17 %) is what should be trusted; this run's "L=6 doesn't fit" finding is specific to the (h=448, heads=7, slices=112) layout, not to depth.

### Suggested follow-ups

1. **Re-run L=6 with power-of-2 layout: hidden=448, heads=8 (head_dim=56), slices=128.** This is the cheapest test of the kernel-layout hypothesis — should cut epoch time by 30–50 %. If it lands near 25 min/epoch, the PR's budget thesis is recoverable.
2. **Or hidden=384, heads=8, slices=96 → head_dim=48.** Bigger param drop (~9 M) but fully aligned. Trades capacity for throughput — informative on whether the perf gap is layout-bound vs param-bound.
3. **Don't pursue h=448/heads=7 further** — it is strictly dominated by both PR #666 and the SOTA baseline.
4. **Profile a single forward pass under nsys / torch profiler** before another full run, to confirm the kernel-layout suspicion concretely.
5. **Stay with PR #666's L=6/h=512 layout if budget allows** — its EP3=7.17 % is the most credible "does deeper help?" signal we have.

### Peak memory

≈ 84 GB / 98 GB per GPU (system metric `system.gpu.*.memory ≈ 86 %`).

### W&B run ID

`b6gc6tgd` (group `yi-round39-depth-l6-h448`, run name `thorfinn/l6-hidden448-heads7`).

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #682 because ADVISOR: Closing as a negative result. The non-power-of-2 attention layout (heads=7, slices=112) triggered a severe CUDA/FlashAttention slow path on H100s — W&B confirms epoch time of ~146 min vs the expected ~20 min (≈7× regression). With only 2 epochs completed in the budget, best val_abupt=16.88% vs baseline 7.39%. The budget-fit hypothesis is rejected for this layout.

Key lesson: On H100s, always use power-of-2 head/slice counts (8/128, 4/64, etc.) to stay in the fast kernel path. If you want to fit within a smaller parameter budget while keeping heads=8/slices=128, reduce hidden_dim to 416 or 448 and accept the slight rounding (448 with heads=8, head_dim=56 is non-standard but slice count is still power of 2 if slices=128).

Thank you for the careful investigation — this is a useful empirical data point..
